python/lsst/dax/obscore/siav2.py

```diff
@@ -0,0 +1,280 @@
+                        )
+                    )
+            else:
+                # Can include all instruments in query, although
+                # binding does not work.
+                instrs = ",".join(repr(inst) for inst in instruments)
+                instrument_wheres.append(WhereBind(where=f"instrument IN ({instrs})", bind={}))
+        elif "band" in dims and "bands" in siav2_bind:
+            # Band is not needed for an instrument query since
+            # we will be using physical filters for those.
+            wheres.append(WhereBind(where="band IN (bands)", bind={"bands": siav2_bind["bands"]}))
+        if pos:
+            extra_dims = set()
+            region_dim_element = dims.region_dim
+            if not region_dim_element:
+                if "exposure" in dims:
+                    # Special case for Rubin default universe.
+                    region_dim = "visit_detector_region"
+                    extra_dims = {"visit"}
+                else:
+                    _LOG.warning("Can not support POS query for dataset type %s", dataset_type_name)
+                    del cfg.dataset_types[dataset_type_name]
+                    continue
+            else:
+                region_dim = region_dim_element.name
+            wheres.append(
+                WhereBind(
+                    where=f"{region_dim}.region OVERLAPS(region)",
+                    bind={"region": siav2_bind["region"]},
+                    extra_dims=extra_dims,
+                )
+            )
+        if time:
+            time_dim = dims.timespan_dim
+            if not time_dim:
+                _LOG.warning("Can not support TIME query for dataset type %s", dataset_type_name)
+                del cfg.dataset_types[dataset_type_name]
+                continue
+            wheres.append(WhereBind(where=f"{time_dim}.timespan OVERLAPS(ts)", bind={"ts": siav2_bind["ts"]}))
+        if exptime:
+            if "exposure" in dims:
+                exp_dim = "exposure"
+            elif "visit" in dims:
+                exp_dim = "visit"
+            else:
+                _LOG.warning("Can not support EXPTIME query for dataset type %s", dataset_type_name)
+                del cfg.dataset_types[dataset_type_name]
+                continue
+            ranges = [float(e) for e in exptime.split()]
+            if math.isfinite(ranges[0]):
+                wheres.append(WhereBind(where=f"{exp_dim}.exposure_time > {ranges[0]}", bind={}))
+            if math.isfinite(ranges[1]):
+                wheres.append(WhereBind(where=f"{exp_dim}.exposure_time < {ranges[1]}", bind={}))
+
+        if instrument_wheres:
+            where_clauses = [WhereBind.combine([iwhere] + wheres) for iwhere in instrument_wheres]
+        else:
+            where_clauses = [WhereBind.combine(wheres)]
+
+        dataset_type_wheres[dataset_type_name] = where_clauses
+
+    return dataset_type_wheres
+
+
+def siav2_query(
```

Comment:
@stvoutsin does this API work for you as a start? Where the strings are exactly what SIAv2 wants so `CIRCLE 20 2 0.5` (POS) and `55456.0 +Inf` (TIME) and `400e-9 600e-9` (BAND).

Reply:
For the first point I had no idea that someone could specify multiple values of all the parameters.  With the way I've written it I can't currently support OR and it will require some thought. You want `Iterable[str]` for all of the parameters in this API?

MAXREC should be straightforward.

Reply:
Any thought on `query_siav2` vs `siav2_query`?

Reply:
Just found the bit of text in the standard:

> All query parameters are multi-valued which means multiple occurrence s of the parameter=value pairs as specified in the DALI recommendation [1] are permitted. The constraints from multiple occurrences of a parameter are combined with a logical OR operator. The constraints from different parameters are combined with a logical AND operator.

Reply:
> Any thought on `query_siav2` vs `siav2_query`?

I think I'd lean towards `siav2_query` (protocol first then mechanism) unless there are existing conventions within the project that do it the other way around

Reply:
We are both parsing all the parameters? Seems wasteful. I'm also validating the parameters in a pydantic model and I pretty much have to do that to support the command-line testing interface. I'm happy to do the ORing -- I just have a higher priority ticket today that needs finishing before I get back to this.

Reply:
I could skip the validation on my end, but the FastAPI /query route has to handle all possible parameters in the SIAv2 spec (2.1 {query} resource section in spec) (some missing from the siav2_query because they are not applicable I imagine). 

There are a couple questions if I do so:
- Will siav2_query raise a validation error which i will catch and produce the VOTable error response (3.2 Errors section in spec). Or will it produce a VOTable with the error message as the response?

- What to do when I get a param that is not supported on your end (for example "specrp", "fov") and it's not formed correctly? I think the expectation is to return an error message (UsageFault error) that this query param is not formed correctly?

Reply:
Currently my pydantic model raises with a validation error. I'm not producing a VOTable error response and it would seem a bit odd for the python API to do that. Sounds like we have to duplicate all the validation code, or share a pydantic model. I'm expecting you to pass me the strings (for POS, BAND etc) as you received them rather than giving me a `SIAV2Parameters` model but is there any way we can share and extend that pydantic model and you pass that to my API?

Currently my API doesn't even accept fov or specrp so if you try it they will fail. I could change my API to accept everything but what should I do if I get something I don't know how to handle? I am also wondering what I'm supposed to do with warnings I issue.

Reply:
Regarding the validation exception being raised I agree that works for me. We can then catch that and form it into a VOTable on the FastAPI app's end, and modify things like the status code of the response accordingly.
I also think we can probably share the pydantic model and avoid duplication. And also yes I assume we can also send you the `SIAV2Parameters` as input rather than individual strings, which might slightly make extensions to support additional params easier.

In terms of the additional params not accepted in the current API, could we add the full list in the pydantic `SIAV2Parameters` model and just validate & ignore them in the actual query mechanism? This way I don't need to do any validation on those additional params. Of course this is only assuming this is the expected behaviour of the service. If we're ok ignoring query params that are formatted wrong and not producing a validation error, then I could just ignore them completely and just send along the `SIAV2Parameters` (or strings) that are supported.

Are the warnings relevant to the user and if so should these warning appear in the result VOTable? In this case do we want to encapsulate the results with the warnings in a new `SIAV2Results` class?

Reply:
Sounds great. I'm happy to add everything to `SIAv2Parameters` and validate and then ignore ones I don't care about for the moment.

The warnings are for things like specifying a timespan for something that has no timespan, or specifying a region in the query when the dataset type that is requested is not spatial. I think whether the warning is useful depends on how much SIAv2 knows about dataset types -- if the dataset types always come from configuration then the warnings are going to be more confusing than anything. If a user can specify that they only want biases but they give a region then we have to decide whether we return no results and a reason or raise an error saying that POS is never going to work for a bias... This is something we can finesse.